File: src/DFS+BFS 필수 문제/2644_촌수계산.py
n = int(input())
p1, p2 = map(int, input().split())
m = int(input())
graph = [[] for _ in range(n + 1)]
visited = [False] * (n + 1)

for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

result = []


def dfs(v, cnt):
    cnt += 1
    visited[v] = True
    if v == p2:
        result.append(cnt)
    for i in graph[v]:
        if not visited[i]:
            dfs(i, cnt)


cnt = 0
dfs(p1, cnt)
if len(result) == 0:
    print(-1)
else:
    print(result[0] - 1)

File: src/DFS+BFS 필수 문제/2644_촌수계산.cpp
/*

bfs를 이용한 cnt:

visited 배열에 cnt를 저장

*/
#include<iostream>
#include<vector>
#include<queue>
#define MAX 101
using namespace std;


int bfs(int s, int d, vector<int> graph[], int visited[])
{   
    queue<int> q;
    q.push(s);

    while(!q.empty())
    {
        int x = q.front();
        q.pop();

        for(int i=0; i<graph[x].size(); i++)
        {
            int nx = graph[x][i];
            if(!visited[nx])
            {
                q.push(nx);
                visited[nx] = visited[x]+1;
            }
            if(nx == d)
            {
                cout << visited[nx] << endl;
                exit(0);
            }
        }
    }

    return -1;
}
int main()
{
    int n, p1, p2, m;
    cin >> n >> p1 >> p2 >> m;

    int visited[MAX] = {0, };
    vector<int> graph[MAX];
    for(int i=0; i<m; i++)
    {
        int u,v;
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    bfs(p1, p2,graph, visited);
    cout << -1 << endl;


}